src/compiler/interpreter.jl

```diff
@@ -212,4 +212,34 @@ let # overload `inlining_policy`
     end
 end
 
-end # module Interpreter
+import Core.Compiler: abstract_call, abstract_call_known, ArgInfo, StmtInfo, AbsIntState, get_max_methods,
+                      CallMeta, Effects, NoCallInfo, widenconst, mapany
+
+struct AutodiffCallInfo <: CallInfo
+    # ...
+    info::CallInfo
+end
+
+function abstract_call_known(interp::EnzymeInterpreter, @nospecialize(f),
+        arginfo::ArgInfo, si::StmtInfo, sv::AbsIntState,
+        max_methods::Int = get_max_methods(interp, f, sv))
+
+    (; fargs, argtypes) = arginfo
+
+    if f === Enzyme.autodiff && length(argtypes) >= 4
+        if widenconst(argtypes[2]) <: Enzyme.Mode && widenconst(argtypes[3]) <: Enzyme.Annotation && widenconst(argtypes[4]) <: Type{<:Enzyme.Annotation}
+          arginfo2 = ArgInfo(
+            fargs isa Nothing ? nothing : [:(Enzyme.autodiff_deferred), fargs[2:end]...],
```

Comment:
You must actually change the IR, I believe.
A simple version would be to just add to the Overlay Table a "autodiff" -> "autodiff_defereed"

Reply:
Yeah, I agree.
```julia
Base.Experimental.@MethodTable ENZYME_TABLE
Base.Experimental.@overlay ENZYME_TABLE function autodiff(...)
    [... the implementation for the deferred autodiff ...]
end
Core.Compiler.method_table(interp::EnzymeInterpreter) = 
    Core.Compiler.OverlayMethodTable(get_inference_world(interp), ENZYME_TABLE)
```
This allows `EnzymeInterpreter` to use the defferred `autodiff` implementation instead of the usual `autodiff` implementation automatically. I think that would work, since when an user calls `autodiff(...)` it dispatches to the usual `autodiff` implementation, which then kicks off the entire Enzyme compilation.

Reply:
so one issue here is that we already receive a method table from GPUCompiler [e.g. for CUDA/etc] 

Also I agree methodtable is cleaner, but the current code here does work [as also confirmed by CI]

Reply:
I am still surprised that this works... You are presenting conflicting information to the abstract interpreter. the CallInfo says one thing, and the IR says another. This may currently "work", but you might also encounter a situation where the code uses the IR as the source of truth.

That's why in my various attempts at something similar, during inlining I replace the IR with something else.